### PR TITLE
Noise Filter scale

### PR DESCRIPTION
**Pull request by garrettwrong**
This is a draft of one of a few possible patches to address #496  (some notes there).

I'm drafting it for discussion at our weekly dev meeting.

### CODE REVIEW

**Comment by garrettwrong** [comment]
I changed this based on our dev discussion.  This solution leaves Cov2D and the noise filter alone. Instead we assume the general case is that we will need/want to recompute the noise var immediate before cov2d.

If a user _really_ thinks they know their noise var (or wants to force 1 etc), it will still allow for that.

Also patches the example experiment.  Ran small case. Looked okay.  Running larger case, then will undraft.

**Comment by garrettwrong** [comment]
Tested with large data set (10028, 360x360, 10k images).

Also tested the CLI `aspire denoise` which has similar function to the experiment. Had same issues.

Should be good to go.

**Review by chris-langfield — DISMISSED** [review]

**Comment by garrettwrong** [comment]
@chris-langfield , I think the three of us (Yunpeng, you, myself) were in agreement that there was a bug in the Aniso variance.  I looked at it a little bit more and am hoping this patch fixes it.  I also added an additional test.

Maybe we should wait for @janden to also review the change before merging, but you should do another review as well. Thanks.

File: tests/test_anisotropic_noise.py
Comment (chris-langfield) [inline tests/test_anisotropic_noise.py]: Parseval?

File: tests/test_anisotropic_noise.py
Comment (chris-langfield) [inline tests/test_anisotropic_noise.py:136]: frequency coefficients

File: tests/test_anisotropic_noise.py
Comment (chris-langfield) [inline tests/test_anisotropic_noise.py]: Parseval I'm 99% sure 😛

**Review by chris-langfield — CHANGES_REQUESTED** [review]

File: tests/test_anisotropic_noise.py
Comment (garrettwrong) [reply tests/test_anisotropic_noise.py]: yes, you are right, I was rushing. Will change the spelling mistakes. Thanks

**Review by chris-langfield — APPROVED** [review]
Nice!

File: src/aspire/commands/denoise.py
Comment (janden) [inline src/aspire/commands/denoise.py:73]: Just out of curiosity, when would this happen?

**Review by janden — APPROVED** [review]
This looks good, but doesn't fix the issue in the batched version, so maybe we should make an issue there. A problem there will be to structure the estimation so that the relevant quantities are estimated in parallel, which is not trivial. The alternative would be to homebrew the variance estimation there, but it's not very satisfactory.

File: src/aspire/commands/denoise.py
Comment (garrettwrong) [reply src/aspire/commands/denoise.py:73]: When a user asks to downsample to a resolution that is greater than the source data.

The way the code was written previously max_resolution was used, which could be anything, but it was never clamped to min(max_resolution, src.L).   This is just a verbose way of doing that.

**Comment by garrettwrong** [comment]
> This looks good, but doesn't fix the issue in the batched version, so maybe we should make an issue there. A problem there will be to structure the estimation so that the relevant quantities are estimated in parallel, which is not trivial. The alternative would be to homebrew the variance estimation there, but it's not very satisfactory.

I don't understand. You are suggesting we need to compute the white noise estimate per batch (not wholesale)?

**Comment by janden** [comment]
> I don't understand. You are suggesting we need to compute the white noise estimate per batch (not wholesale)?

No it's just that we have two methods for covariance estimation, the regular one and the batched one. This fixes the former but not the latter. The variance estimation still takes place over the whole dataset, but it needs to be adapted for the batched covariance estimation to ensure we only perform one pass through the dataset.

**Comment by garrettwrong** [comment]
This does nothing to change the number of passes through the data in either case... FWIW, the noise estimation is batched in both cases.

**Comment by janden** [comment]
> This does nothing to change the number of passes through the data in either case...

Right. Before the user would have to run the variance estimation before calling the cov2d code (assuming they didn't just assume that it was one).

> FWIW, the noise estimation is batched in both cases.

I'm not following. This PR does not touch `BatchedRotCov2D`, which still requires `noise_var` as an input to `get_covar` (currently defaults to zero). This will also have to be fixed but done in a way that doesn't add an extra pass over the data.

**Comment by garrettwrong** [comment]
Okay sure, I understand what you mean now.  You are talking about an optimization.  In this issue I was just trying to fix outputting junk (potatoes) from our example experiment and CLI for _calling_ cov2d... when you said "fixed" I thought you were implying that was not resolved; I was pretty concerned.

For that optimization I don't think we should take away from other efforts. I'll make an issue in the morning. It is an easy win we can cover in the future.

**Comment by garrettwrong** [comment]
(by batched I just mean that the NoiseEstimator was consuming the data in batches, but this was not related to what you were talking about)

**Pull request merged**
